CAIPs/caip-x.md

```diff
@@ -0,0 +1,339 @@
+
+3. **Wallet Initiated Permissions Revocation:**
+
+   - **Current Method:** "If a respondent (e.g. a wallet) needs to initiate a new session, whether due to user input, security policy, or session expiry reasons, it can simply generate a new session identifier to signal this notification to the calling provider."
+   - Given this language it is unclear if a wallet can revoke permissions without creating a new session. It is also therefore unclear if a wallet can revoke some subset of permissions without creating a new session.
+   - **Proposed Method:** Wallet publishes and caller/dapp listens for an event `providerAuthorizationChanged` with the new full sessionScope.
+
+4. **Dapp Initiated Permissions Revocation:**
+
+   - **Current Method:** "if a caller [dapp] needs to initiate a new session, it can do so by sending a new request without a sessionIdentifier."
+   - **Proposed Method:** Use `provider_revoke` to revoke specific permissions or the entire existing session.
+
+5. **Retrieving Current Permissions:**
+
+   - **Current Method:** Track permissions externally or infer from session creation.
+   - **Proposed Method:** Use `provider_getSession` to retrieve the current permissions of the session.
+
+## Specification
+
+### New Methods
+
+#### `provider_augment` // open to a different name here!
+
+Adds new permissions to an existing session.
+
+**Parameters:**
+
+- `sessionId` (string, optional): The session identifier.
+- `scopes` (object, required): An object containing the new scopes to be added, formatted according to CAIP-217.
+
+**Initial Session Scopes:**
+
+```json
+{
+  "eip155:1": {
+    "methods": ["eth_signTransaction"],
+    "notifications": ["accountsChanged"],
+    "accounts": ["eip155:1:0xabc123"]
+  },
+  "eip155:137": {
+    "methods": ["eth_sendTransaction"],
+    "notifications": ["chainChanged"],
+    "accounts": ["eip155:137:0xdef456"]
+  }
+}
+```
+
+**Example Request:**
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "method": "provider_augment",
+  "params": {
+    "scopes": {
+      "eip155:1": {
+        "methods": ["eth_sendTransaction"],
+        "notifications": ["chainChanged"]
+      }
+    }
+  }
+}
+```
+
+**Example Response:**
+
+```json
+{
+  "id": 1,
+  "jsonrpc": "2.0",
+  "result": {
+    "success": true,
+    "updatedScopes": {
+      "eip155:1": {
+        "methods": ["eth_signTransaction", "eth_sendTransaction"],
+        "notifications": ["accountsChanged", "chainChanged"],
+        "accounts": ["eip155:1:0xabc123"]
+      },
+      "eip155:137": {
+        "methods": ["eth_sendTransaction"],
+        "notifications": ["chainChanged"],
+        "accounts": ["eip155:137:0xdef456"]
+      }
+    }
+  }
+}
+```
+
+**Explanation:**
+
+- The `provider_augment` method adds the specified scopes to the current session's permissions. If the scope already exists, the new methods and notifications are merged with the existing ones.
```

Comment:
Why `updatedScopes` and not `sessionScopes`?

I'd specify more concretely, i.e.:

- if wallet rejects the augmentation, is the expected behavior returning the status-quo scopes or returning nothing or returning an explicit `no.` of some kind?

Reply:
`sessionScopes` works for me. I will also beef up the specification to include error codes and rejection flows.